update sgl-kernel for w4afp8_machete

mianpeng opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation
the sgl-kernel part for https://github.com/sgl-project/sglang/pull/8573

This PR  supports a quantized offline model [DeepSeek-R1-AWQ-W4AFP8](https://huggingface.co/TMElyralab/DeepSeek-R1-AWQ-W4AFP8) .
We accomplished the W4AFP8 implementation for both MoE and Linear Method，by optimizing and tuning the Machete gemm kernel . The result simultaneously boost sustained throughput and cut time latency on a single H20 or H100.

For more details, please refer to the code repository [lyraW4AFp8](https://github.com/TMElyralab/lyraW4AFp8.git)


## Model

We offered W4AFP8 AWQ quantized models :

- [DeepSeek-R1-AWQ-W4AFP8](https://huggingface.co/TMElyralab/DeepSeek-R1-AWQ-W4AFP8)
- [DeepSeek-R1-0528-AWQ-W4AFP8](https://huggingface.co/TMElyralab/DeepSeek-R1-0528-AWQ-W4AFP8)
- [DeepSeek-V3-0324-AWQ-W4AFP8](https://huggingface.co/TMElyralab/DeepSeek-V3-0324-AWQ-W4AFP8)
- [DeepSeek-V3.1-AWQ-W4AFP8](https://huggingface.co/TMElyralab/DeepSeek-V3.1-AWQ-W4AFP8)


## Citation
We are TMElyralab, the Acceleration Team from Tencent Music Entertainment (TME).

``` bibtex
@Misc{TMElyralab_2025,
  author =       {Sa Xiao, Mian Peng, Haoxiong Su, Kangjian Wu, Bin Wu, Yibo Lu, Qiwen Mao, Wenjiang Zhou},
  howpublished = {\url{https://github.com/TMElyralab}},
  year =         {2025}
}
```

mianpeng committed:
update sgl-kernel w4afp8_machete

mianpeng committed:
update format

mianpeng committed:
fix format bug

mianpeng commented:
@FlamingoPg @HydraQYH  Could you please review this PR? If you have a moment. If there are any problems, I’ll fix them right away. Thank you!
This PR is the sgl-kernel parts splited from https://github.com/sgl-project/sglang/pull/8573
And the moe&linear python parts is https://github.com/sgl-project/sglang/pull/9737

HydraQYH reviewed (commented):
This is a great job. Can you add unit tests and benchmarks for these kernels?

CarsonSo commented on sgl-kernel/csrc/quantization/extensions/dispatch_policy.hpp:
```diff
@@ -0,0 +1,36 @@
+//  `ScaleGranularityM` indicates that scaling granularity is
+//  `size<0>(TileShape_MNK{})` along M.
+template <int ScaleGranularityM = 0>
+struct KernelTmaWarpSpecializedCooperativeFP8BlockScaledSubGroupMAccum : KernelTmaWarpSpecializedCooperative {};
```
SubGroupM a reserved param for FP8 block-wise scaling factor along M dimension. It stands for number or contiguous rows assigned to each warp-group. Each tile will be divided into #sub-groups = TileM / SubGroupM along M.

mianpeng committed:
update test&bench

mianpeng committed:
Merge branch 'main' into lyra_w4afp8_sglkernel

mianpeng commented:
> This is a great job. Can you add unit tests and benchmarks for these kernels?

@HydraQYH Thanks! I have added the unit tests and benchmarks

mianpeng commented on sgl-kernel/csrc/quantization/machete/machete_mm_kernel.cuh:
```diff
@@ -0,0 +1,342 @@
+    // {Accum, C, C_layout, D, D}
+    EpilogueArguments epilogue_arguments{};
+
+    // 创建尾操作参数：处理 w channel-quant 反量化 + a token-group-wise 反量化
```
Ok. I changed the annotation to English

mianpeng commented:
> It seems that some CI failed. Can you take a look?

Thanks. I will fix these two days.
Do we need to run PR TEST on amd and ascend NPU? However, we don't have these NPU machines.

mianpeng committed:
reset utils in sglang python

mianpeng committed:
reset utils in sglang python

mianpeng committed:
Merge branch main into lyra_w4afp8_sglkernel

mianpeng commented:
> LGTM~ btw need to fix conflicts

@HydraQYH @FlamingoPg The conflicts is fixed .  Thanks.

mianpeng commented:
@HydraQYH Could this PR be merged? I only made some changes in sgl-kernel/python/sgl_kernel/__init__.py to fix conflicts
<img width="696" height="178" alt="image" src="https://github.com/user-attachments/assets/f67a68a1-748c-4708-b0b2-d11739250c1f" />

HydraQYH reviewed (approved):
LGTM

HydraQYH commented:
@FlamingoPg It seems I can't restart CI without the run-ci tag. Can you take a look?

FlamingoPg commented:
No other problem, btw there are two more points that need to be changed:

- The file sgl-kernel/include/scalar_type.hpp actually already exists. You made a duplicate copy.
- In the cpp/cuda files, can the namespace be changed from vllm to sglang? Please keep the “adapted” and “copied” comments as they are.

After that we can merge this~ cc: @mianpeng

mianpeng committed:
update spacename & code clean

mianpeng committed:
Merge branch 'main' into lyra_w4afp8_sglkernel

mianpeng commented:
> > No other problem, btw there are two more points that need to be changed:
> > 
> > * The file sgl-kernel/include/scalar_type.hpp actually already exists. You made a duplicate copy.
> > * In the cpp/cuda files, can the namespace be changed from vllm to sglang? Please keep the “adapted” and “copied” comments as they are.
> > 
> > After that we can merge this~ cc: @mianpeng
> 
> @FlamingoPg Please review the code. I deleted sgl-kernel/csrc/quantization/extensions/scalar_type.hpp and changed 'vllm' to 'sglang'

@FlamingoPg Can this PR be merged? Thanks.

mianpeng committed:
clean CMakeLists

mianpeng commented:
> > Sorry for the late reply. I’ll continue the review now. @mianpeng
> > @FlamingoPg Thanks! Would it be possible to merge this PR now? Is there anything else I need to do?
> 
> Pass the CI, it could be merged

@FlamingoPg  I've noticed 3 CI checks failed, maybe because the branch is out of date. Do I need to update this branch with the lateset sglang code?

FlamingoPg commented:
> > > Sorry for the late reply. I’ll continue the review now. @mianpeng
> > > @FlamingoPg Thanks! Would it be possible to merge this PR now? Is there anything else I need to do?
> > 
> > 
> > Pass the CI, it could be merged
> 
> @FlamingoPg I've noticed 3 CI checks failed, maybe because the branch is out of date. Do I need to update this branch with the lateset sglang code?

Yes, rebase master plz.

mianpeng committed:
Merge branch 'main' into lyra_w4afp8_sglkernel

mianpeng commented:
> @mianpeng Wait `pr-test-finish` this CI passed

@FlamingoPg This PR only adds some code to the sgl-kernel. From the logs below, the failures appear to be unrelated to this PR. Could you please skip these CI tests? These CI jobs have already been running for quite a while. Thank you.

<img width="483" height="365" alt="企业微信截图_5dd9433b-a03d-48e9-bae6-424550cf9a21" src="https://github.com/user-attachments/assets/6e37b985-33bb-48bd-b778-24587926b243" />
unit-test-backend-1-gpu
<img width="1068" height="797" alt="企业微信截图_c236672c-6630-4222-9391-e29af500400d" src="https://github.com/user-attachments/assets/a3f05efa-733a-481c-ae7e-ab3b0f4278a6" />
unit-test-backend-4-gpu
<img width="1202" height="346" alt="企业微信截图_41b96779-4e3f-4a32-9de0-cc8cb8993d79" src="https://github.com/user-attachments/assets/cc83a3a4-be7b-4179-9d94-77bbe05ce808" />
performance-test-2-gpu
<img width="775" height="138" alt="企业微信截图_40235ef3-13d4-44b9-9d89-1271a1bb1539" src="https://github.com/user-attachments/assets/225f89f0-7048-400c-ab62-339b026afff8" />

FlamingoPg commented:
And these CI fixed in recent PR, I cant help you rebase main @mianpeng

mianpeng committed:
Merge branch 'main' into lyra_w4afp8_sglkernel

FlamingoPg commented:
> > And these CI fixed in recent PR, I cant help you rebase main @mianpeng
> 
> @FlamingoPg I've just merged main into this branch. Could you please take a look when you have a moment? Thank you!

I see, no problem

mianpeng committed:
Merge branch 'main' into lyra_w4afp8_sglkernel

FlamingoPg commented:
@mianpeng Sorry to told you that some CI fixes from yesterday caused conflicts. Could you rebase onto master? Once the CI passes, I’ll help merge it.

mianpeng committed:
Merge branch 'main' into lyra_w4afp8_sglkernel

mianpeng committed:
Merge branch 'main' into lyra_w4afp8_sglkernel

mianpeng commented:
@HydraQYH This PR has now been open for over two months. Because the sglang codebase is changing frequently, the CI never stabilizes. We keep merging updates and rerunning the checks, which has taken a lot of time and effort on both sides. Is there anything I can do to resolve this besides continuing to wait? I’m happy to do in any way that would make the process smoother.

mianpeng committed:
Merge branch 'main' into lyra_w4afp8_sglkernel

HydraQYH commented:
> @HydraQYH This PR has now been open for over two months. Because the sglang codebase is changing frequently, the CI never stabilizes. We keep merging updates and rerunning the checks, which has taken a lot of time and effort on both sides. Is there anything I can do to resolve this besides continuing to wait? I’m happy to do in any way that would make the process smoother.

@FlamingoPg I noticed that we've both approved this PR. Do you have any suggestions regarding the questions @mianpeng  raised? I haven't followed this PR for a while, so I'm not sure why there hasn't been a merge yet.

mianpeng committed:
Merge branch 'main' into lyra_w4afp8_sglkernel

mianpeng committed:
Merge branch 'main' into lyra_w4afp8_sglkernel

FlamingoPg commented:
@HydraQYH I don't have any other issues, it's just that the code updates in the quantization area are quite frequent. Also, the CI wasn't very stable a while ago, and I need all GPU CI tests to pass before I can merge with merge permissions.

FlamingoPg commented:
@mianpeng Perhaps you could also add permissions for me and @HydraQYH  to your repository, so we can help you with updates as well.

HydraQYH committed:
Merge branch 'main' into lyra_w4afp8_sglkernel
